provider/resource.go

```diff
@@ -301,37 +307,38 @@ func arrayValue(vars resource.PropertyMap, prop resource.PropertyKey, envs []str
 
+	// Default environment to Public unless specified.
+	environment := environments.AzurePublic()
 	envName := stringValue(vars, "environment", []string{"ARM_ENVIRONMENT"})
-	if envName == "" {
-		envName = "public"
+	switch cases.Title(language.English).String(envName) {
+	case environments.AzureChinaCloud:
+		environment = environments.AzureChina()
+	case environments.AzureUSGovernmentCloud:
+		environment = environments.AzureUSGovernment()
 	}
 
 	//check for auxiliary tenants
 	auxTenants := arrayValue(vars, "auxiliaryTenantIDs", []string{"ARM_AUXILIARY_TENANT_IDS"})
 
 	// validate the azure config
-	// make a Builder
-	builder := &authentication.Builder{
-		SubscriptionID:       stringValue(vars, "subscriptionId", []string{"ARM_SUBSCRIPTION_ID"}),
-		ClientID:             stringValue(vars, "clientId", []string{"ARM_CLIENT_ID"}),
-		ClientSecret:         stringValue(vars, "clientSecret", []string{"ARM_CLIENT_SECRET"}),
-		TenantID:             stringValue(vars, "tenantId", []string{"ARM_TENANT_ID"}),
-		Environment:          envName,
-		ClientCertPath:       stringValue(vars, "clientCertificatePath", []string{"ARM_CLIENT_CERTIFICATE_PATH"}),
-		ClientCertPassword:   stringValue(vars, "clientCertificatePassword", []string{"ARM_CLIENT_CERTIFICATE_PASSWORD"}),
-		MsiEndpoint:          stringValue(vars, "msiEndpoint", []string{"ARM_MSI_ENDPOINT"}),
-		AuxiliaryTenantIDs:   auxTenants,
-		ClientSecretDocsLink: "https://www.pulumi.com/docs/intro/cloud-providers/azure/setup/#service-principal-authentication",
+	credentials := auth.Credentials{
+		ClientID:                      stringValue(vars, "clientId", []string{"ARM_CLIENT_ID"}),
+		ClientSecret:                  stringValue(vars, "clientSecret", []string{"ARM_CLIENT_SECRET"}),
+		TenantID:                      stringValue(vars, "tenantId", []string{"ARM_TENANT_ID"}),
+		Environment:                   *environment,
+		ClientCertificatePath:         stringValue(vars, "clientCertificatePath", []string{"ARM_CLIENT_CERTIFICATE_PATH"}),
+		ClientCertificatePassword:     stringValue(vars, "clientCertificatePassword", []string{"ARM_CLIENT_CERTIFICATE_PASSWORD"}),
+		CustomManagedIdentityEndpoint: stringValue(vars, "msiEndpoint", []string{"ARM_MSI_ENDPOINT"}),
+		AuxiliaryTenantIDs:            auxTenants,
 
 		// Feature Toggles
-		SupportsClientCertAuth:         true,
-		SupportsClientSecretAuth:       true,
-		SupportsManagedServiceIdentity: boolValue(vars, "msiEndpoint", []string{"ARM_USE_MSI"}),
-		SupportsAzureCliToken:          true,
-		SupportsAuxiliaryTenants:       len(auxTenants) > 0,
+		EnableAuthenticatingUsingClientCertificate: true,
+		EnableAuthenticatingUsingClientSecret:      true,
+		EnableAuthenticatingUsingManagedIdentity:   boolValue(vars, "msiEndpoint", []string{"ARM_USE_MSI"}),
+		EnableAuthenticatingUsingAzureCLI:          true,
 	}
 
-	_, err := builder.Build()
+	_, err := auth.NewAuthorizerFromCredentials(context.TODO(), credentials, environment.MicrosoftGraph)
```

Comment:
[nit] `context.Background()` should be fine here, authentication is a long-running background thing